libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -1404,45 +1425,85 @@ void IRGeneratorForStatements::appendExternalFunctionCall(
+				add(<pos>, 4)
+			</bareCall>
+			<argumentString>
+		)
 
 		let <success> := <call>(<gas>, <address>, <?hasValue> <value>, </hasValue> <pos>, sub(<end>, <pos>), <pos>, <reservedReturnSize>)
-		<?noTryCall>
+		<?revertOnFailure>
 			if iszero(<success>) { <forwardingRevert>() }
-		</noTryCall>
+		</revertOnFailure>
 		<?hasRetVars> let <retVars> </hasRetVars>
-		if <success> {
-			<?dynamicReturnSize>
-				// copy dynamic return data out
-				returndatacopy(<pos>, 0, returndatasize())
-			</dynamicReturnSize>
-
-			// update freeMemoryPointer according to dynamic return size
-			mstore(<freeMemoryPointer>, add(<pos>, <roundUp>(<returnSize>)))
-
-			// decode return parameters from external try-call into retVars
-			<?hasRetVars> <retVars> := </hasRetVars> <abiDecode>(<pos>, add(<pos>, <returnSize>))
-		}
+		<?returnSuccessAndReturndata>
+			<?returnReturndata> let <returndataVar> := <extractReturndataFunction>() </returnReturndata>
+		<!returnSuccessAndReturndata>
+			if <success> {
+				<?dynamicReturnSize>
+					// copy dynamic return data out
+					returndatacopy(<pos>, 0, returndatasize())
+				</dynamicReturnSize>
+
+				// update freeMemoryPointer according to dynamic return size
+				mstore(<freeMemoryPointer>, add(<pos>, <roundUp>(<returnSize>)))
+
+				// decode return parameters from external try-call into retVars
+				<?hasRetVars> <retVars> := </hasRetVars> <abiDecode>(<pos>, add(<pos>, <returnSize>))
+			}
+		</returnSuccessAndReturndata>
 	)");
 	templ("pos", m_context.newYulVariable());
 	templ("end", m_context.newYulVariable());
+	templ("bareCall", funType.isBareCall());
+	templ("returnSuccessConditionAndReturndata", returnSuccessConditionAndReturndata);
 	if (_functionCall.annotation().tryCall)
 		templ("success", m_context.trySuccessConditionVariable(_functionCall));
+	else if (returnSuccessConditionAndReturndata)
+	{
+		if (funKind == FunctionType::Kind::Send)
+		{
+			templ("success", IRVariable(_functionCall).name());
+			templ("returnReturndata", false);
+		}
+		else
+		{
+			templ("success", IRVariable(_functionCall).tupleComponent(0).name());
+			templ("returnReturndata", true);
+			templ("returndataVar", IRVariable(_functionCall).tupleComponent(1).name());
+			templ("extractReturndataFunction", m_utils.extractReturndataFunction());
+		}
+	}
 	else
 		templ("success", m_context.newYulVariable());
 	templ("freeMemory", freeMemory());
 	templ("shl28", m_utils.shiftLeftFunction(8 * (32 - 4)));
-	templ("funId", IRVariable(_functionCall.expression()).part("functionIdentifier").name());
-	templ("address", IRVariable(_functionCall.expression()).part("address").name());
+
+	if (!funType.isBareCall())
+		templ("funId", IRVariable(_functionCall.expression()).part("functionIdentifier").name());
+
+	if (funKind == FunctionType::Kind::ECRecover)
+		templ("address", "1");
+	else if (funKind == FunctionType::Kind::SHA256)
+		templ("address", "2");
+	else if (funKind == FunctionType::Kind::RIPEMD160)
+		templ("address", "3");
```

Comment:
These system contract addresses are hardcoded in other places too, e.g. in `ExpressionCompiler.cpp`:

https://github.com/ethereum/solidity/blob/a371910674108c7798930d0f20dfe7830471e43c/libsolidity/codegen/ExpressionCompiler.cpp#L908-L917

Maybe it would be better to have constants for them? Makes it harder to mix them up or even just make a typo.